### DISCUSSION TITLE
add another store-dir config as a fallback

### DISCUSSION THREAD

**Discussion by futurist** [pnpm/pnpm#15907] [Ideas]
- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

**Describe the user story**

Our use case is mount `.pnpm-big` from a center common storage (Same device), but it's read-only, and we set the env-var `npm_config_store_dir` to this folder. But since it's read-only, run `pnpm i` cannot write files into it and may cause problems.

**Describe the solution you'd like**

Maybe we can set `.pnpm-store` (env-var/cli option controllable) as a fallback dir if written to `.pnpm-big` failed?

And then we can sync the R/W folder `.pnpm-store` back into `.pnpm-big` for later use.

**Comment by zkochan** [comment]
If you have the ability to sync the R/W folder with the `.pnpm-big` folder, why don't you just sync it prior to installation and use just the regular `.pnpm-store`?

**Comment by futurist** [comment]
> If you have the ability to sync the R/W folder with the `.pnpm-big` folder, why don't you just sync it prior to installation and use just the regular `.pnpm-store`?

It's a bit difficulty to do, since:
1. the R/O `.pnpm-big` really very big like 50GB, R/W `.pnpm-store` can be small like 100MB (just for one repo size)
2. the sync from R/W to R/O is totally optional and may have other ways like run `pnpm store add <pkgs>` from the center node(has R/W perm for `.pnpm-big`) only to enrich the center store, but it's only during night maybe.
3. the main point is we cannot mount `.pnpm-big` as R/W since it's shared by everyone, and we don't want people delete it, it's totally OK if we found a way to make pnpm work happy with this R/O big store. Another R/W store and the syncing are additional things that can be omitted maybe.

**Comment by zkochan** [comment]
I would suggest to use Vedaccio instead of the mounted readonly store.

We could add such feature to pnpm but it would not be a small change, and this does not seem like a frequent scenario.

**Comment by thynson** [comment]
I don't understand what's your scenario. 
Maybe you are trying to boost the CI process? As another solution, in this case, I would suggest using a layered filesystem, just like what docker did, to hold the store.

**Comment by futurist** [comment]
It's in WebIDE, and we want people install deps ASAP, with a big R/O `.pnpm-big` store folder, but cannot continue due to this feature issued.

**Comment by zkochan** [comment]
At this point

https://github.com/pnpm/pnpm/blob/bcc82cd685ee1772eb95dae85d475f33d0bbc1b6/packages/package-requester/src/packageRequester.ts#L383-L387

if the package is not found in the content-addressable store, we may search in a read-only CAFS. But we will still need to verify the contents of files in that readonly store and then copy the files.

What if instead of a readonly store, there will be a readonly cache of package tarballs? This would be easier to implement. We'd only have to update the [tarball-fetcher](https://github.com/pnpm/pnpm/tree/main/packages/tarball-fetcher) to look in the readonly cache before fetching packages from the registry.

**Comment by futurist** [comment]
@zkochan It's very exciting if this feature can be implemented, and will be a killing feature vs other npm solutions, since it's  enterprise friendly.

**Comment by futurist** [comment]
> At this point
> 
> https://github.com/pnpm/pnpm/blob/bcc82cd685ee1772eb95dae85d475f33d0bbc1b6/packages/package-requester/src/packageRequester.ts#L383-L387
> 
> if the package is not found in the content-addressable store, we may search in a read-only CAFS. But we will still need to verify the contents of files in that readonly store and then copy the files.
> 
> What if instead of a readonly store, there will be a readonly cache of package tarballs? This would be easier to implement. We'd only have to update the [tarball-fetcher](https://github.com/pnpm/pnpm/tree/main/packages/tarball-fetcher) to look in the readonly cache before fetching packages from the registry.

Besides the tarball thing, it's great if the side effects can also be in the read-only store, since we found the `side-effects-cache` options very useful, the installation time significantly reduced due to the cache.

**Comment by zkochan** [comment]
I think we can try to implement this as suggested initially.

**Comment by wiruzman** [comment]
Any updates?

**Comment by altaywtf** [comment]
I'm working on a fix for this.

---
Written by an agent (Claude Code, Opus 5.5).

**Comment by altaywtf** [comment]
A fix is ready for review in https://github.com/pnpm/pnpm/pull/15861. CI is green.

Written by an agent (Claude Code, Opus 5.5).
